### DISCUSSION TITLE
GHA visualization graph considered harmful

### DISCUSSION THREAD

**Discussion by magicus** [community/community#18035] [Actions]
**TL;DR:** The job dependency visualization graph box provides no value, but actively hinders you from getting to the relevant data. It's counterproductive and should be removed, hidden by default or, as a last resort, moved to the absolute bottom of the page).

**Long story:**

The visualization graph box, on the "Summary" page of a workflow run, is problematic from a UI/UX perspective. See for instance this run: https://github.com/magicus/jdk/actions/runs/2442545082. The gray `main.yml` box occupies most of my laptop screen. I need to scroll past it to be able to see the **Annotations**, **Artifacts** and markdown summary sections. These sections are far more important to me, yet they are unavailable to me by default.

To make matters worse, scrolling is almost impossible on a device with a touch pad. If your cursor is over the job list to the left, it only scrolls the job list, which does not make the visualization graph scroll off the screen. And if your cursor is over the graph itself, the scrolling is interpreted as a zoom, and scrolling is blocked. So to actually be able to scroll past the graph, you must position your mouse cursor carefully on the thin border outside of the graph.

All of this is annoying, but what makes matter even worse is that the visualization graph is of very little use. Let's go back to my example at https://github.com/magicus/jdk/actions/runs/2442545082. This is a highly complex workflow scenario. But this means that the graph starts out in a zoomed-out mode, where any text is practically unreadable. If you really want to use the graph, you need to zoom in, and pan around. But even if you do this, you will soon discover that the graph lies to you. What seems like edges connecting nodes are, in fact, not always such. A job node can be placed "on top" of an edge, so the edge "passes through", below the node. (A more reasonable expectation is that the graph layout algorithm would have routed the edge around the node.) So to *truly* understand the relationships, you need to click on each node in term, because only then do the proper edges get highlighted in blue. For a great example of this, see: https://github.com/magicus/jdk/actions/runs/360576283.

But, you might say, these examples are very complicated runs, and the graph was not designed to handle them. Well, what then, was it designed for? My second half of this criticism is that for simpler run, the graph is utterly pointless. If you have only a single job, the graph fills no purpose at all. The only real situation I can see is if you have e.g. two separate jobs, like `build` and `test`, where the latter depends on the former. But then again, is there *really* a need to visualize this? For an example of a trivial job, see: https://github.com/Wynntils/Artemis/actions/runs/2453440666. Even here, though it adds absolutely no value, the graph box steals 1/4 of my screen space, and I need to scroll to get to the bottom of the Gradle summary.

I hope you will seriously consider doing something about this. My primary suggestion is to get rid of the graph box, since it does not meet user expectations. If you are unable or unwilling to do that, at the very least try to hide it. Make it "collapsed" by default, and/or move it to the bottom of the summary page, below the annotations, artifacts and markdown summary. (And while you're at it, I think the markdown summary is more useful to have higher up on the page than the artifacts. Typically you'll check the summary before deciding if you need to download artifacts.)

**Comment by mbraunwart** [comment]
To add to this, it would be beneficial if there were a way to use the workflows to customize how the visualizer will display certain jobs or collections of jobs. Now that Reusable Workflows can be up to 4 layers deep, jobs that reside within those workflows can cause the visualizer to expand greatly, especially if building a continuous workflow that deploys to multiple environments in a dependency chain.

It would be great if you could signify in the reusable workflow that the visualizer should group all or some jobs together as matrix jobs are group or something similar.

**Comment by olsen232** [comment]
I raised this issue separately with GitHub before I knew of this discussion. In our case the particular issue was that new users were not finding our artifacts, since they are hidden below the graph and scrolling the page - or indeed learning that it is even scrollable - is extremely difficult for the uninitiated as magicus describes. 
So far Github has closed this as working-as-intended so we've just updated all of our documentation with instructions that will hopefully help new users to scroll the page and find the artifacts.

> #### Description
>
> A new contributor visits the REPO/actions/runs/RUNID and looks for the artifacts.
> The "build.yml" matrices view is taking up much of the screen, so artifacts are not visible - they are down below.
> The contributor tries to scroll down using their touchpad or scroll wheel, but their mouse cursor is above the largest panel - the build.yml matrices view. This captures their scroll-event and prevents the page from scrolling.
> Since the new contributor cannot scroll, they assume the page is only that which they can see right now. They will not be able to find > the artifacts. They will need to start a discussion to get help from existing users who have learnt the esoterics of how to work with GitHub.
>
> #### Reproduction Steps
>
> Visit, for example, https://github.com/koordinates/kart/actions/runs/3888824147
> Position your mouse cursor above the "build.yml" matrices view.
> Try to scroll using the scroll wheel or touchpad.

**Comment by magicus** [comment]
Personally, I have resorted to using an adblocker and am blocking the element containing the visualization graph. But as you say, it does not help newcomers. :-(

**Comment by mikebollandajw** [comment]
how do i disable this? it freezes some of our larger projects and is unecessary

i have added the filter mentioned above by magicus as a work around
! 24 Jun 2024 https://github.com
github.com##.mx-md-1.px-2.position-relative.height-full.width-full

**Comment by markfickett** [comment]
I find the visualization graph useful to understand execution progress, but I would definitely like to put custom Markdown summaries above it.

**Comment by rf-molsson** [comment]
We have 6 consecutive stages/columns in our graph, and during the third stage we have 32 boxes running in parallel. The font size in our graph is so tiny we can't even read the text inside each bubble/node. The UX of the graph is absolute terrible indeed.

**Comment by mohitkumaragarwal** [comment]
Is there any solution found for this issue yet with actual root cause fix?
